### ISSUE TITLE
Slicing longer full-text into text chunks to pre-populate a territory with statements

### ISSUE THREAD

**Issue by davidzbiral** [DISSINET/InkVisitor#1610]
Adam has this idea of implementation: the idea would be to let the user open a dialog where

they can paste a longer free text and select a territory in a suggester
the application will slice the free text into a slices
user can then check the slices, edit them
and after clicking "accept" all the statements will be created in the chosen Territory.

**Comment by adammertel** [comment]
The first mockup by @pondrejk 
![image](https://user-images.githubusercontent.com/12932677/223155823-b8f38784-1883-4bba-8fe5-cc105b325aff.png)

**Comment by davidzbiral** [comment]
@adammertel : this is quite viable I think. Only:
1. I think there is not the need to have the Submit button in the first stage before any parsing, because there is nothing to store.
2. I would have the labels as follows: "Pre-fill text fields", "Clear" (as now), "Segment" (instead of Parse), and "Create statements" (instead of "Submit" - from submit, it is not so completely clear that e.g. 20 statements will be created by this action).

I also tag @GideonK .

**Comment by pondrejk** [comment]
@davidzbiral @adammertel 

Updated mockup 
![image](https://user-images.githubusercontent.com/8287578/223828096-1cbf1692-98fe-4d9a-9ab6-04397f602ca3.png)

Changes:
- it is possible to set a different separator than .
- the segmented lines now form their own text fields that can be edited freely
- refresh button allows additional segmentation, either by placing the original separator somewhere into the sentence or by choosing a different separator
- it is possible to delete lines by x button

**Comment by davidzbiral** [comment]
> the segmented lines now form their own text fields that can be edited freely

Good move. But @pondrejk and @adammertel , the segmentation by a character separator will not work very well - we will need [dependency parsing](https://towardsdatascience.com/natural-language-processing-dependency-parsing-cf094bbbe3f7) (with the help of @GideonK) for this to spare any bigger amount of time.

**Comment by adammertel** [comment]
@davidzbiral I am not sure how much effort is to implement the dependency parsing mechanisms into InkVisitor. Are you at this moment okey with the character separator and manual edit of the statements or do you want to postpone this feature until we can implement the dependency parsing?

**Comment by GideonK** [comment]
I am currently in the process of comparing pre-processing pipelines of Latin to each other, which includes checking the quality of the part-of-speech tags etc. that would inform a parser in the next step and directly affects its quality. There are a few different approaches, but I have found one that seems to be compatible for Latin, at least for the purpose of dependency parsing in InkVisitor:

- CoreNLP is an NLP package written in Java developed by the Stanford NLP Group. The Stanza Python package (previously called StanfordNLP) includes a precompiled version of CoreNLP; so it seems, no need to set up a server (I need to test the proposed pipeline, but Stanza is also implemented by CLTK, the Classical Language Toolkit, which I've already used for experiments).
- Stanza includes [Universal Dependency models](https://stanfordnlp.github.io/stanza/available_models.html) including for Latin.
- There is a tool called [Semgrex](https://aclanthology.org/2023.tlt-1.7.pdf), which implements regex-like patterns to match dependencies. It has support for both Java and Python, the latter of which can also use displaCy as in David's link, for vizualization. So we can possibly use this tool to build and apply rules in helping us decide how to split sentences into statements. Semgrex was reimplemented as part of spaCy 3.0 with a tool called [DependencyMatcher](https://spacy.io/api/dependencymatcher).
- There is an [online interface](http://universal.grew.fr/?corpus=UD_Latin-UDante@2.11) where you search for examples from existing UD treebanks, including Latin.

It may be better to use a sentence boundary detection tool (should be implemented as part of the Stanza pipeline) instead of selecting a specific punctuation mark, as abbreviations will be handled incorrectly, and it may be that some sentences have unusual endings (single-character ellipses, etc.). Quite possibly, we don't even need to use a separate tool, as (from the paper) "Semgrex reads dependency trees from CoNLL-U files or parses dependencies from raw text using the associated CoreNLP parser."

If we can get this running, the next will be building the rules to decide how sentences are split into statements, including examples such as the one provided by David. As I'm both not a Latin expert nor completely informed on the UD tagset and grammar, this may need to be built up from the ground. From a quick look (and in my view a little easier to understand from the DependencyMatcher link), the operators can be quite useful.

**Comment by adammertel** [comment]
@GideonK Thank you for a knowledgeable answer. Unfortunately, we are now using Javascript on both Backend and Frontend; therefore, incorporating a Python package would mean a new level of complexity.

Now I see two options. Either (i) use an existing JS package,  for example, [wink-nlp](https://github.com/winkjs/wink-nlp) or [nnsplit](https://github.com/bminixhofer/nnsplit), but unfortunately, none of them support Latin. Or (ii) use a dumb method of parsing the text by given characters (e.g., `.`, `,`, `;`) and depend on the user to edit the result manually. 

While this function is not going to be used daily, and even the good language models cannot perfectly accommodate CASTEMO principles,  I would go with (ii).

Using `regex` would be ideal while it is possible to incorporate it in any programming language quickly, but I am not sure how production-ready the idea is.

**Comment by GideonK** [comment]
@adammertel This might work as long as there are not many abbreviations in the text, as regex is not sufficient for this otherwise. Some algorithms are more language agnostic so could be looked into. But if we expect just a certain set of characters to look out for, sure (ii) should be fine.

Another option could be to pre-split the sentences with e.g. CLTK before they are being dropped into InkVisitor? i.e. each one appears on a separate line already, and there is no need to in corporate the splitter into the workflow. Or is this contrary to what you are planning?

**Comment by tomaham** [comment]
Well. It should not be too hard to stitch a python api script called as
http service.

**Comment by davidzbiral** [comment]
@adammertel : I do not understand the technicalities. The character separation will not work very well, so then the critical point is the convenience of how the re-segmentation upon request is implemented. 
The solution should take into account the broader horizon which InkVisitor must follow - becoming an annotation tool holding full texts (#1531).

**Comment by davidzbiral** [comment]
Paused because a thorough full-text management is in view, but will be quite needed then.

**Comment by davidzbiral** [comment]
@adammertel After viewing some tens of examples, I can say that we need dependency parsing, and delimiting it by punctuation is not viable at all.

But we have a solution. Adam and @GideonK please meet and work together to implement Hackathon6-developed clause_text delimitation method (i.e. finding spans of text which are candidates for statements based on PoS identifying VERB and then their most distant dependents to the left and to the right in the full-text). This is exactly the same span that we need for this purpose. (And Gideon, we also need for this purpose to delimite the subordinate clause/s spanned by the span, and in the way that you enclose them with curly brackets, for this InkVisitor purpose we need to delimit statement 1 part 1 with statement 1 anchors, then delimit statement 2 (subordinate clause), and then again return with a second pair of anchors to statement 1 (part 2). I hope it is clear, we can all meet to implement this.

**Comment by davidzbiral** [comment]
@adammertel If you think it is a good solution, move to 1.4.1 - segmentor not needed in 1.4.0, conveniently manually editing statement anchors is. As you want.

**Comment by davidzbiral** [comment]
This will still need a more discursive form than issue list - a document to consider, talk, brainstorm. Created [here](https://docs.google.com/document/d/1DhfMr50Q9_arMcCj-g7DdeiGyfu8fryKqeT5sVM3qbs/edit).

**Comment by davidzbiral** [comment]
@adammertel I am also signalling that we have ca. 300 gold-standard segmentation examples which could be used as training data, in Hackathon 11 (16-20 Sep 2024) we should achieve something like 700-1000.

**Comment by davidzbiral** [comment]
@adammertel , @GideonK How this is going? I would hate to postpone this too much, but should it be 1.4.0, or 1.4.1? Are enough human resources available to make it 1.4.0?

**Comment by adammertel** [comment]
@davidzbiral I would move this to 1.4.1 considering that we would like to have the 1.4.0 as soon as possible.

**Comment by GideonK** [comment]
> @adammertel , @GideonK How this is going? I would hate to postpone this too much, but should it be 1.4.0, or 1.4.1? Are enough human resources available to make it 1.4.0 if we are not using punctuation-based but dependency-parsing-based approach (on the basis of clause delimitation done by Gideon)?

Sorry for only getting back to you now, we did discuss this in person and on Slack. To my understanding from your feedback on the performance of the dependency-based statement delimitation in the validation dataframe, it has issues but is workable. As you know, Adam and I had a meeting and I have more or less an idea what needs to happen from my side.

Firstly, the current implementation is embedded in a relation extraction workflow where a sentence is processed, and from left to right we search for verbs. If it is the head of the sentence (main verb), and because of legacy reasons (Agency, such as producing different types of statistics, etc.) it is treated in a separate command as other verbs, but eventually, they are processed the same way. The standard input therefore is a verb and its dependencies, should there be any. Of course it is modular, so the class that actually performs the tagging can be reused.

Although each token is associated with its sentence through the spaCy doc object, basically we only work with the clause when creating the output tags. So whatever the verb is that is the current focus verb, from the point of view of the clause, that is the "main" verb and will govern whatever appears in <S1>. It will therefore not handle any superordinate clauses, so an output can never start with <S2> and then <S1> - rather, the real <S2> will be handled in another run, with the current <S1> as <S2> (i.e. the superordinate clause will start with <S1><S2>... but the current (subordinate) one just with <S1>...).

If I remember correctly, Adam has mentioned that we need to distinctive features: First, we need to be able to produce a continuous stream of text that includes tags. This is a different assignment than what we currently have because we work with focus verbs (each new verb is the current focus) where for each one, we generated a separate <S1>, <S2>, etc. However, in a continuous stream of text, we cannot have multiple S1s, etc. in the output. I therefore need to think of a way to either merge outputs or adapt the algorithm to work with a single sentence possibly with multiple verbs. I'm not clear yet on what to do here.

Secondly, my understanding is that a user should be able to select any piece of text and then the suggester will automatically add suggested tags in the text. At the least there should also be some significant adaptations. Firstly, it is a bad idea to try and parse a string less than a full sentence, so probably, we need to train a doc object from the context. Then when a text is selected, we will have access to the perceived subtree spans that exist within that selection. It might become problematic if we go over sentence boundaries, but we should of course have a default action for each possible user decision. I think it's easiest if the tags are made beforehand in the background, and after any editing, redone on the new selection.

**Comment by davidzbiral** [comment]
@GideonK 
> However, in a continuous stream of text, we cannot have multiple S1s, etc. in the output.
Yes - of course this will not be S1, but the numbering will be continuous from S1 to S9999999999999999999, so a new main clause will not be S1 again. S1 will not repeat in the output that the parser will return. It will be identifiable as main clause still by the fact that it is not enclosed in any other tag.

**Comment by davidzbiral** [comment]
David and Katalin do not think it will massively speed up CASTEMO.
Rather important for MGSTEMO? 
Moving to 1.4.3.
